inc/class-wpseo-endpoint-factory.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Internals
+ */
+
+/**
+ * Class WPSEO_Endpoint_Factory
+ */
+class WPSEO_Endpoint_Factory {
+
+	/**
+	 * @var array The valid HTTP methods.
+	 */
+	private $valid_http_methods = array(
+		'GET',
+		'PATCH',
+		'POST',
+		'PUT',
+		'DELETE',
+	);
+
+	/**
+	 * @var array The arguments.
+	 */
+	protected $args = array();
+
+	/**
+	 * @var string The namespace.
+	 */
+	private $namespace;
+
+	/**
+	 * @var string The endpoint URL.
+	 */
+	private $endpoint;
+
+	/**
+	 * @var callable The callback to execute if the endpoint is called.
+	 */
+	private $callback;
+
+	/**
+	 * @var callable The permission callback to execute to determine permissions.
+	 */
+	private $permission_callback;
+
+	/**
+	 * @var string The HTTP method to use.
+	 */
+	private $method;
+
+	/**
+	 * WPSEO_Endpoint_Factory constructor.
+	 *
+	 * @param string   $namespace           The endpoint's namespace.
+	 * @param string   $endpoint            The endpoint's URL.
+	 * @param callable $callback            The callback function to execute.
+	 * @param callable $permission_callback The permission callback to execute to determine permissions.
+	 * @param string   $method              The HTTP method to use. Defaults to GET.
+	 *
+	 * @throws WPSEO_Invalid_Argument_Exception The invalid argument exception.
+	 */
+	public function __construct( $namespace, $endpoint, $callback, $permission_callback, $method = 'GET' ) {
```

Comment:
Minor, but we might wanna use `WP_REST_Server::READABLE` instead of hard-coding 'GET'.